Imaging Browser: Add DICOM download: Redmine 12580

MounaSafiHarab opened the pull request:
Projects can download the .tar that includes raw (DICOM) images, as well as .log and .meta (some meta data related to the insertion progress)

1) One .tar per candidate per visit:
![downloaddicom](https://user-images.githubusercontent.com/15198379/36400381-8dcd732a-159e-11e8-9638-62003b6b083e.png)

2) Two .tar per candidate (example here illustrated becaus2 the candidate had to repeat a specific sequence another day, but under the same visit):
![updatedimagbrowserscreenshottwotars](https://user-images.githubusercontent.com/15198379/36542794-e19b225a-17af-11e8-94f9-49f368833f43.png)

christinerogers commented on modules/imaging_browser/test/imaging_browser_test_plan.md:
```diff
@@ -11,7 +11,7 @@
 ### B. ViewSession / Volume List
 
-8. Sidebar:  all links work 
+8. Sidebar:  all links work, inlcuding the Download DICOM option
```
- typo: inlcuding -> including
- If there are multiple tars, they should be listed (and download on each should work)

MounaSafiHarab replied:
@christinerogers 
1) typo fixed! thanks
2) yes, this should work; I tested it as we had a case in CCNA wwhich had its initial MRi visit split into two scanning sessions and the changes here work as well. I fixed the screenshot at the beginning of the pull request to show this case.

PapillonMcGill commented on modules/imaging_browser/php/imaging_session_controlpanel.class.inc:
```diff
@@ -96,13 +96,20 @@ class Imaging_Session_ControlPanel
+        foreach ($qresult as $k=>$v) {
+            $tarchiveID = $v['TarchiveID'];
+            $tarIDToTarLoc[$tarchiveID] =$v['ArchiveLocation'];
+        }
```
suggest using array_values() instead of foreach. Prevent unused variable $k.

MounaSafiHarab replied:
removed unused `$k`. thanks!

davidblader commented:
needs rebase

MounaSafiHarab commented:
> needs rebase

I just did now that PR#3500 is merged.

cmadjar commented on modules/imaging_browser/README.md:
```diff
@@ -78,7 +78,10 @@ mantis_url - This setting defines a URL for LORIS to include a link to for bug r
   `$LORIS/tools/CouchDB_Import_MRI.php` alongside all other CouchDB
   import scripts, but should logically be considered part of this module.)
 - The imaging browser module includes links to BrainBrowser to visualize data.
+- The control panel on the viewsession page includes links to instruments
```
I would tend to right "View Session" page

MounaSafiHarab replied:
done, changed in in the 4 occurrences

cmadjar commented on modules/imaging_browser/README.md:
```diff
@@ -78,7 +78,10 @@ mantis_url - This setting defines a URL for LORIS to include a link to for bug r
   as configured by the study.
-- The control panel on the viewsession page includes links to the DICOM Archive
-  for any DICOM tars associated with the given session.
+- The control panel on the viewsession page includes links to the DICOM Archive 
```
Same here, "View Session" instead of viewsession

MounaSafiHarab replied:
done

MounaSafiHarab commented:
> @driusan, @MounaSafiHarab could it be possible to send a warning that the file cannot be found instead of the error page? It would look cleaner.

perfectly valid point, but just to be clear, this is the same error you get if you try to download a MINC file that does not exist (see below).
![getfileminc](https://user-images.githubusercontent.com/15198379/36995041-faebf16a-2080-11e8-92ff-7595dae4bc01.png)

it is in the `get_file`, in the section that handles the  file non-existent condition:
`if (!file_exists($FullPath)) {
    error_log("ERROR: File $File does not exist");
    header("HTTP/1.1 404 Not Found");
    exit(5);
}
`

So not sure I should change it here, or this is the subject of another pull request....

cmadjar commented:
@MounaSafiHarab I agree with you, it should probably be a separate ticket. Do you want to create a redmine ticket for it? In the meantime, I'll approve your PR :). Thanks!

MounaSafiHarab commented:
> @MounaSafiHarab I agree with you, it should probably be a separate ticket. Do you want to create a redmine ticket for it? In the meantime, I'll approve your PR :). Thanks!

Done, Redmine #14029

driusan commented on htdocs/mri/jiv/get_file.php:
```diff
@@ -92,6 +94,12 @@
 
+// If $File contains "DCM_", prefix automatically inserted by the
+// LORIS-MRI pipeline, identify it as $FileExt: "DICOMTAR"
+if (strpos($File, "DCM_") ) {
```
missing `=== 0`. The comment should also point out that it's a hack, since there's no such thing as a DICOMTAR extension.

MounaSafiHarab replied:
@driusan 
you can not assume the `DCM` is at the beginning of the filename. It used to be the case, but not anymore (now the file stored within a year directory, followed by the filename which then starts with `DCM_`), so in the tarchive table, a typical ArchiveLocation is:
`2016/DCM_2016-09-13_ImagingUpload-9-50-abCD77.tar`

driusan replied:
In that case, it still needs `!== false` to differentiate from finding at the 0 index vs not finding in the string at all.

MounaSafiHarab replied:
thanks, done!

ridz1208 commented:
@kongtiaowang too many changes after you tested, please test again

driusan commented on modules/imaging_browser/php/imaging_session_controlpanel.class.inc:
```diff
@@ -105,13 +105,20 @@ class Imaging_Session_ControlPanel
+        $tarIDToTarLoc =array();
+        foreach ($qresult as $v) {
+            $tarchiveID = $v['TarchiveID'];
+            $tarIDToTarLoc[$tarchiveID] =$v['ArchiveLocation'];
```
I'm surprised this passed PHPCS

MounaSafiHarab replied:
I don't really know;
I added an extra space after the `=` sign, so it is a manual change; and this passes phpcs if this was your main concern (I still do not know how it passed before)
not sure if the other issue is also the line before where the `==` is not aligned with the line underneath it, but when I did force it to manually be aligned with the line below, it actually gave me phpcs error!

MounaSafiHarab commented:
@nicolasbrossard 

I addressed your comments on @davidblader PR (#2967) here for the `get_file.php`
which incidentally could also address one of @driusan's earlier comments on this pull request (for checking the beginning of the filename starting with `DCM` now that I added the basename).

MounaSafiHarab commented:
I rebased this on minor since PR #2967 has been merged, so I do not need the `get_file` anymore...

is there anything else this PR is waiting on?
@davidblader @driusan @PapillonMcGill @cmadjar ?
